### DISCUSSION TITLE
Pipe By Pipe By Get Doc Object For Review Pipe Outcome ?

### DISCUSSION THREAD

**Discussion by kamlesh0606** [explosion/spaCy#12986] [Help: Coding & Implementations] [answered]
Hello Team,

Can Get Doc Object at Pipe execution Over For review Doc Outcome ?

Use Cases,
1) When add 7-9 Custom Pipe that time, training Team needs to get Pipe to Pipe Outcome and understand problem at which pipe have ?
2) Tracking Dataset and Rules Parts of Pattern understand  
3) When Apply `measurement_merger`  Pipe apply 2-3 Time For get more useful outcome. that at time , Doc Object Need to understand Data Flow and other things. And also write Pattern on the new Token at next leave Pipe.
4) Manage Multiple level Object need for parsing word at final Logic with help of all Level Doc Object And other Details..

Team, If you need example dataset of above points , Please update me.

Thanks
Kamlesh.

**Comment by danieldk** [comment, marked as answer]
We currently don't have any callbacks on pipe completion, but I can think of two ways to go about. First, you could make a loop calling the pipes similar to how `Language` does:

```
for name, proc in self.pipeline:
    doc = proc(doc)
    # Add code here to inspect doc
```

You shouldn't use this in production code, since `Language` does many additional things, but it can be ok for debugging. The other option is to [make a custom pipe component](https://spacy.io/usage/processing-pipelines#custom-components). This component could just return the doc as-is, but you could insert your instrumentation code into it. You could then use this component after the pipe(s) that you want to instrument.

**Reply by kamlesh0606** [reply]
@danieldk  Thank you for given update.

Pipe to Pipe Data Save inside Doc->user_data attribute ? 

Save Time , `to_json()` of Doc and Save in `user_data`.  is it ok ?

Or Any other good way for it ?

**Reply by danieldk** [reply]
If you want to keep track of the annotations in the document, so that they are serialized with the document, storing them in `user_data` sounds like a reasonable approach. But putting a JSON-serialized representation of the whole document would be a bit unwieldy (would make the resulting docs very large).
